crates/ra_hir_ty/src/infer/pat.rs

```diff
@@ -185,6 +185,22 @@ impl<'a, D: HirDatabase> InferenceContext<'a, D> {
                 self.write_pat_ty(pat, bound_ty);
                 return inner_ty;
             }
+            Pat::Slice { prefix, slice, suffix } => {
+                if let Ty::Apply(ApplicationTy { ctor: TypeCtor::Slice, parameters }) = expected {
```

Comment:
I think we should rather use the expected type as a *hint*; we still want to call `infer_pat` on all subpatterns regardless of expected type. See here how we extract an expected element type for array expressions:
https://github.com/rust-analyzer/rust-analyzer/blob/ea67e2346e64a1a14206e6d0f163c6c2e19d0afb/crates/ra_hir_ty/src/infer/expr.rs#L472-L477
Then we can just call `infer_pat` on all the prefix and suffix patterns with that expected type, and always return the slice type as the type of the pattern (even if we don't have an expected type and don't know the element type, we know that the whole pattern needs to be some slice).

Reply:
Hey, thanks for the feedback. I think I'm still too new to this codebase to follow exactly what you are suggesting, but I think if there was a test case that highlights the situation you are concerned about I could probably figure it out. 

Does creating a new test with `let slice: &[_] = &[];` provide a good example of the situation you are concerned about?

Reply:
I was concerned about a situation like
```rust
fn make_anything<T>() -> T { loop {} }
let arr = &make_anything();
match arr {
    &[1u32, x] => x,
}
```
and was assuming that rustc would figure out the type of `x` and `arr` there, but apparently it doesn't. The reason that it doesn't is that slice patterns can also match arrays, though. So we also need to handle an array type in `expected.ty`, e.g.:
```rust
let arr = [1, 2];
match arr {
    [1u32, b] => b,
}
```

But also, I just think the logic here is just more complicated than it needs to be. We don't need to match on `(prefix.as_slice(), slice, suffix.as_slice())`, we can just do the for loop on `prefix` (and `suffix`). And we don't need to wrap everything in the `if let`.

Reply:
See my latest commit, which removes the match statement and handles prefix and suffix. 

I'm not sure how I could get rid of the `if let` since that is how we get `parameters`? 

Also note in that latest commit that this changed the type inference of a different pattern test. Is this expected/desired?

Let me know if I am going the wrong direction here and I can revert this commit, I just felt like maybe having a concrete code change would facilitate discussion.

Reply:
That is the right direction, I think :slightly_smiling_face: 

> I'm not sure how I could get rid of the if let since that is how we get parameters?

If we don't have a slice (or array), we can just create a type variable for the element type. That way, we can still infer e.g. `x` in the pattern `[1u32, x]` (the type variable will get unified with the type of the first pattern, so by the time we get to the second pattern, we'll know the type). That's why I was referencing this code:
https://github.com/rust-analyzer/rust-analyzer/blob/ea67e2346e64a1a14206e6d0f163c6c2e19d0afb/crates/ra_hir_ty/src/infer/expr.rs#L472-L477
It's pretty close to what you have (it's just using a macro to make the patterns a bit shorter, and it's matching slices and arrays).

> Also note in that latest commit that this changed the type inference of a different pattern test. Is this expected/desired?

Yes, I think the change there is correct.

One more thing: If we're matching an array, we should return the array type. For that, I'd check whether the expected type is an array or slice, and return the corresponding type in the end. If it's neither, I think it's fine to just fall back to slice.

Reply:
Thanks for your patience in guiding me through this. My latest commit adds an array test case, and handles them in a way which I believe to be correct. 

The match statement doesn't feel super elegant here, but it isn't terrible I guess. Let me know if you are aware of a cleaner way to express that.

Reply:
I think the match statement is fine!